recipe/2024-06-private-bicep-registries.md

```diff
@@ -0,0 +1,258 @@
+| -------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| OCI compliant registry | An OCI-compliant registry refers to a container registry that adheres to the specifications set by the Open Container Initiative  |
+| ORAS | ORAS is the de facto tool for working with OCI Artifacts.(https://oras.land/) |
+| Private Bicep Registry | A private Bicep registry is a dedicated repository for storing and managing Bicep modules which is not accessible publicly need to authenticate into it for registry operations.| 
+
+## Objectives
+> **Issue Reference:** https://github.com/radius-project/radius/issues/6917
+
+### Goals
+- Add support to publish bicep recipes to a private bicep registry.
+- Enable support to use bicep recipes stored in private bicep registries.
+- Support all OCI compliant private registries.
+
+### Non goals
+
+- Workload identity based authentication(Will be covered as part of different design) 
+- Configure private OCI registry credentials in Radius via CLI and API.
+
+### User scenarios (optional)
+As an operator I am responsible for maintaining Bicep recipes for use with Radius. Bicep recipes used contains sensitive information, proprietary configurations, or data that should not be shared publicly and its intended for internal use within our organization and have granular control over who can access, contribute to, or modify bicep recipes. So I store the recipes in a private bicep registry. And I would like to use these private registries as template-path while registering a bicep recipe.
+
+
+## User Experience (if applicable)
+**Sample Input:**
+Command to publish recipe to private bicep registry.
+```diff
+-rad bicep publish --file ./redis-test.bicep --target br:ghcr.io/myregistry/redis-test:v1
+
++rad bicep publish --file ./redis-test.bicep --target br:ghcr.io/myregistry/redis-test:v1 --username <user> --password <password>
+```
+
+**Sample Output:**
+
+No change in the output
+
+
+## Design
+### Design details
+Currently, OCI-compliant registries are used to store Bicep recipes, with the ORAS client facilitating operations like publish and pull these recipes from the registries. ORAS provides package auth, enabling secure client authentication to remote registries. Private Registry credentials information i.e username and password must be stored in a Application.Core/secretStores resource and the secret resource ID is added to the recipe configuration.
+
+During the recipe deployment bicep driver checks for secrets information for that container registry the recipe deploying is stored on,  and creates an oras auth client to authenticate the private bicep registries.
+
```

Comment:
It sounds like there are different parts of Radius that will need to authenticate. 

- The CLI (`rad bicep publish`)
- The Bicep recipe driver 

Do these have different requirements? or the same requirements?

Reply:
Yes its different, for the `recipe driver` we expect the users to have bicep recipe configured with `secretstore` information holding private registry credentials. But for `rad bicep publish` we expect username and password as command line args.